test/unit/batch_alloc.c

```diff
@@ -0,0 +1,96 @@
+#include "test/jemalloc_test.h"
+
+#define BATCH ((1U << 16) + 101)
+
+void
+verify_batch(void **ptrs, size_t batch, size_t usize, bool zero) {
+	tsdn_t *tsdn = tsdn_fetch();
+	szind_t ind = sz_size2index(usize);
+	const bin_info_t *bin_info = &bin_infos[ind];
+	const unsigned nregs = bin_info->nregs;
+	void *ref = calloc(1, usize);
+	assert_ptr_not_null(ref, "");
+	for (size_t i = 0, j = 0; i < batch; ++i, ++j) {
+		if (j == nregs) {
+			j = 0;
+		}
+		void *p = ptrs[i];
+		expect_zu_eq(isalloc(tsdn, p), usize, "");
+		if (zero) {
+			expect_d_eq(memcmp(p, ref, usize), 0, "");
+		}
+		free(p);
+		if (j == 0) {
+			expect_true(prof_sample_aligned(p), "");
+			continue;
+		}
+		assert(i != 0);
+		size_t d = (size_t)((uintptr_t)p - (uintptr_t)(ptrs[i - 1]));
+		if (config_prof && opt_prof) {
+			/*
+			 * There are three possibilities:
+			 *
+			 * (a) Both the previous allocation and p are not
+			 *     prof-sampled: they'll surely be adjacent.
+			 * (b) The previous allocation is not prof-sampled and
+			 *     p is prof-sampled: they may or may not be
+			 *     adjacent, and p will be page-aligned.
+			 * (c) The previous allocation is prof-sampled: p will
+			 *     surely not be adjacent to the previous
+			 *     allocation, regardless of whether p is
+			 *     prof-sampled or not, and both will be
+			 *     page-aligned.
+			 *
+			 * Therefore, when p is adjacent to the previous
+			 * allocation, we only know that the previous
+			 * allocation is not prof-sampled, but we know nothing
+			 * about p.  However, if p is not adjacent to the
+			 * previous allocation, then we know that at least one
+			 * of them is prof-sampled, and that p is page-aligned.
+			 */
+			assert_zu_ge(d, usize, "");
+			if (d > usize) {
+				expect_true(prof_sample_aligned(p), "");
+				j = 0;
+			}
+		} else {
+			expect_zu_eq(d, usize, "");
+		}
+	}
+}
+
+TEST_BEGIN(test_batch_alloc) {
+	void *ptrs[BATCH];
+	const size_t size = 7;
+	size_t filled = arena_batch_alloc(ptrs, BATCH, size, 0);
+	assert_zu_eq(filled, BATCH, "");
+	verify_batch(ptrs, BATCH, sz_s2u(size), false);
+}
+TEST_END
+
+TEST_BEGIN(test_batch_alloc_zero) {
+	void *ptrs[BATCH];
+	const size_t size = 7;
+	size_t filled = arena_batch_alloc(ptrs, BATCH, size, MALLOCX_ZERO);
+	assert_zu_eq(filled, BATCH, "");
+	verify_batch(ptrs, BATCH, sz_s2u(size), true);
+}
+TEST_END
+
+TEST_BEGIN(test_batch_alloc_aligned) {
+	void *ptrs[BATCH];
+	const size_t size = 7;
+	const size_t a = 16;
+	size_t filled = arena_batch_alloc(ptrs, BATCH, size, MALLOCX_ALIGN(a));
+	assert_zu_eq(filled, BATCH, "");
+	verify_batch(ptrs, BATCH, sz_sa2u(size, a), false);
+}
+TEST_END
+
+int
+main(void) {
```

Comment:
We should also have tests of stats behavior under calls to batch_alloc (things like malloc counts, nslabs, etc.).

Reply:
And probably zeroing behavior, too.

Reply:
Oh no, the existing test looks good. I guess I managed to forget about it in the 1 minute gap between those two comments.